services/storage/src/simcore_service_storage/api/rest/_simcore_s3.py

```diff
@@ -0,0 +1,160 @@
+import asyncio
+import logging
+from typing import Annotated, Any, cast
+
+from fastapi import APIRouter, Depends, FastAPI, Request
+from models_library.api_schemas_long_running_tasks.base import TaskProgress
+from models_library.api_schemas_long_running_tasks.tasks import TaskGet
+from models_library.api_schemas_storage import FileMetaDataGet, FoldersBody
+from models_library.generics import Envelope
+from models_library.projects import ProjectID
+from servicelib.aiohttp import status
+from servicelib.fastapi.long_running_tasks._dependencies import get_tasks_manager
+from servicelib.logging_utils import log_context
+from servicelib.long_running_tasks._task import start_task
+from settings_library.s3 import S3Settings
+from yarl import URL
+
+from ...dsm import get_dsm_provider
+from ...models import (
+    DeleteFolderQueryParams,
+    FileMetaData,
+    SearchFilesQueryParams,
+    StorageQueryParamsBase,
+)
+from ...modules import sts
+from ...simcore_s3_dsm import SimcoreS3DataManager
+
+_logger = logging.getLogger(__name__)
+
+router = APIRouter(
+    tags=[
+        "simcore-s3",
+    ],
+)
+
+
+@router.post("/simcore-s3:access", response_model=Envelope[S3Settings])
+async def get_or_create_temporary_s3_access(
+    query_params: Annotated[StorageQueryParamsBase, Depends()],
+    request: Request,
+):
+    # NOTE: the name of the method is not accurate, these are not temporary at all
+    # it returns the credentials of the s3 backend!
+    s3_settings: S3Settings = await sts.get_or_create_temporary_token_for_user(
+        request.app, query_params.user_id
+    )
+    return Envelope[S3Settings](data=s3_settings)
+
+
+async def _copy_folders_from_project(
+    progress: TaskProgress,
+    app: FastAPI,
+    query_params: StorageQueryParamsBase,
+    body: FoldersBody,
+) -> Envelope[dict[str, Any]]:
+    dsm = cast(
+        SimcoreS3DataManager,
+        get_dsm_provider(app).get(SimcoreS3DataManager.get_location_id()),
+    )
+    with log_context(
+        _logger,
+        logging.INFO,
+        msg=f"copying {body.source['uuid']} -> {body.destination['uuid']}",
+    ):
+        await dsm.deep_copy_project_simcore_s3(
+            query_params.user_id,
+            body.source,
+            body.destination,
+            body.nodes_map,
+            task_progress=progress,
+        )
+
+    return Envelope[dict[str, Any]](data=body.destination)
+
+
+@router.post(
+    "/simcore-s3/folders",
+    response_model=Envelope[TaskGet],
+    status_code=status.HTTP_202_ACCEPTED,
+)
+async def copy_folders_from_project(
+    query_params: Annotated[StorageQueryParamsBase, Depends()],
+    body: FoldersBody,
+    request: Request,
+):
+    task_id = None
+    try:
+        task_id = start_task(
```

Comment:
Similar to what you did with the files API, you can use a `BaseDeferredHandler` to ensure it ensure the copy is restarted in case of a service restart. This way you do not care if someone restart the service at all.